File: golang/leetcode/41-first-missing-positive/main.go
package main

import (
	"fmt"

	"golang/leetcode/41-first-missing-positive/canonical"
	fixedarray "golang/leetcode/41-first-missing-positive/fixed-array"
)

func main() {
	fmt.Println("fixed-array:")
	{
		nums := []int{1, 2, 0}
		fmt.Println(fixedarray.FirstMissingPositive(nums))
	}
	{
		nums := []int{3, 4, -1, 1}
		fmt.Println(fixedarray.FirstMissingPositive(nums))
	}
	{
		nums := []int{7, 8, 9, 11, 12}
		fmt.Println(fixedarray.FirstMissingPositive(nums))
	}

	fmt.Println("canonical:")
	{
		nums := []int{1, 2, 0}
		fmt.Println(canonical.FirstMissingPositive(nums))
	}
	{
		nums := []int{3, 4, -1, 1}
		fmt.Println(canonical.FirstMissingPositive(nums))
	}
	{
		nums := []int{7, 8, 9, 11, 12}
		fmt.Println(canonical.FirstMissingPositive(nums))
	}
}

File: cpp/leetcode/41-first-missing-positive/main.cpp
// https://leetcode.com/problems/first-missing-positive/

#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int firstMissingPositive(vector<int> &nums)
    {
        for (int i = 0; i < nums.size(); ++i)
        {
            while (true)
            {
                const auto val = nums[i]; // алиас
                if (val <= 0)
                {
                    break;
                }

                if (val > nums.size())
                { // 5 будет на позиции 5-1, а значит ОК
                    break;
                }

                if (nums[val - 1] == val)
                {
                    break;
                }

                swap(nums[i], nums[val - 1]);
            }
        }

        for (int i = 1; i <= nums.size(); ++i)
        {
            if (nums[i - 1] != i)
            {
                return i;
            }
        }

        return nums.size() + 1;
    }
};

int main()
{
    Solution s;
    {
        vector<int> nums = {3, 4, -1, 1};
        cout << s.firstMissingPositive(nums) << endl;
    }
    {
        vector<int> nums = {1, 2, 0};
        cout << s.firstMissingPositive(nums) << endl;
    }
    {
        vector<int> nums = {7, 8, 9, 11, 12};
        cout << s.firstMissingPositive(nums) << endl;
    }
    return 0;
}
